README.rst
=========
Provision
=========

Introduction
============

Provision enables users to deploy customized nodes, either via shell
commands, or as a Python library.  Building on Apache Libcloud, which
provides a common API for various providers, Provision allows users to
specify, for a new node, which files get installed, which scripts get
run, and which public keys have access, in a flexible yet repeatable
way.

In addition to basic configuration decisions (such as disk image,
size, provider, location, and name), Provision supports four
conceptually distinct components that determine how a node gets
deployed.

#. Public keys that grant their respective private keys root access on
   the deployed node.

#. Files that get copied into their specified location on the node.

#. Script templates supporting run time variable substitution that get
   run on the node.

#. Minimal Python configuration code for setting secret keys, creating
   named bundles of scripts and files for specific functionality, etc.

Provision applications are themselves automatically configured from
the provision/defaults directory and optional, additional
configuration directories which themselves contain an __init__.py
file, and possible subdirectories of files, scripts and public keys.

These additional configuration directories can be used to change the
default configuration parameters in any way, and are typically used to
set secret API keys, access keys, and define site-specific
functionality bundles.

A Bundle is a named collection of files and scripts that will get
installed and run on the deployed node.  You can specify a default set
of Bundle names that will get installed for every new node, as well as
specifying which additional bundles to install via command line or
library interface.


Dependencies
============

See setup.py for details on version requirements.  The following
modules will be automatically installed if not present.

* apache-libcloud
* paramiko
* pycrypto
* argparse
* cloudfiles (optional)


Installation
============

Provision uses distribute.setuptools for installation and test running.

To install, cd to top level directory and run::

   $ python setup.py install

To run tests, cd to top level directory and run::

   $ python setup.py test


Usage
=====

Shell Commands
--------------

The following shell commands are available after installation:

* deploy-node
    Typical usage: $ deploy-node -c ~/secrets_dir -b dev
  
* list-nodes
    Typical usage: $ list-nodes -c ~/secrets_dir
  
* destroy-node
    Typical usage: $ destroy-node -c ~/secrets_dir nodename

Several command line arguments are common to all three commands:

* -c --config-paths
    Specify path to a configuration directory. Can be used multiple times.

* -p --provider
    Specify provider (defaults to config.DEFAULT_PROVIDER)

* -u --userid
    Specify user id (defaults to config.DEFAULT_USERID)

* -k --secret_key
    Specify secret key (defaults to config.DEFAULT_SECRET_KEY)

Other command line arguments are specific to individual commands

deploy-node
^^^^^^^^^^^

* -b --bundles
    Specify names of bundles to install.  Can be used multiple times.

* -d --description-file
    If specified, output node description in json format to file.

* -i --image
    Specify image name (defaults to config.DEFAULT_IMAGE_NAME)

* -l --location
    Specify location id (defaults to config.DEFAULT_LOCATION_ID)

* -n --name
     Generate randomized name based on prefix if name not specified.

* -s --size
    Specify node size (defaults to config.DEFAULT_SIZE_ID)

* -t --subvars
     Key=value pairs of template substitution variables. Can be used multiple times.

* -x --prefix
    Use prefix to generate randomized name (defaults to config.DEFAULT_NAME_PREFIX)

destroy-node
^^^^^^^^^^^^

* name
    The name of the node to destroy

* -t --testresults
    Only destroy node if all tests passed in specified junit-style XML formatted file

Configuration Directory Structure
---------------------------------

Provision is not particulary useful out of the box.  At the minimum,
you will need to specify which provider, user id, and secret key to
use to access your account.  This can all be done on the command line,
but it's can be simpler to create a local configuration directory and
either specify its location on the command line, or put it in a
default location that provision will try to load on startup.

Aside from authentication, a configuration directory can be use to
define bundles of associated files and scripts that will get run when
a node is deployed.  It can also read and write any variable defined
in the provision.config module, which gives great flexibility in
determining how the program will act by default.

Provision configuration directories all share the same structure.  At
the top level is a __init__.py file, which gets imported and its
init() function executed during configuration time.

Also at the top level are three directories called "pubkeys",
"scripts", and "files".  Provision uses libcloud, which uses public
key cryptography by default to communicate with the new node.  During
a deploy, it will by default look for the file ~/.ssh/id_rsa.pub and
insert it into the node's /root/.ssh/authorized_keys file.  If it
exists and contains files, provision will also include those public
keys in the new node's authorized_keys.

From the other two directories, files and scripts get loaded into
memory, and are mapped into bundles in __init__.py, which can then be
specified in the command line using -b bundle-name, or added to
DEFAULT_BUNDLES, to get installed for every deploy.

It is sometimes useful to be able to substitute variables into scripts
at runtime.  This can be done by using the --subvars command line
option with script templating.

Embed one of the following lines in a script to activate variable
substitution::

    # provision-template-type: format-string
    or
    # provision-template-type: template-string

See `format string documentation
<http://docs.python.org/library/string.html#format-string-syntax>`_
and `template strings documentation
<http://docs.python.org/library/string.html#template-strings>`_ for
the respective syntaxes.  Also see test cases in
test_script_templates.py.

The __init__.py file can also be used to override default settings in
the provision.config module, which gets passed into init() as a
parameter.

This is an example of an __init__.py file::

    def init(config):
        config.DEFAULT_PROVIDER = 'rackspace'
        config.DEFAULT_USERID = 'user1'
        config.DEFAULT_SECRET_KEY = 'somehardtoguesssecret'

        config.DEFAULT_BOOTSTRAP_BUNDLES.extend(['tz', 'snmpd']
        config.DEFAULT_BUNDLES.extend(['security'])

        config.add_bundle('dev', ['emacs.sh', 'screen.sh'],
                          ['/root/.emacs.d/init.el', '/root/.screenrc'])


For this example, the files directory contains init.el, which will get
installed at /root/emacs.d/init.el in the deployed node, and .screenrc
which gets installed and /root/.screenrc.

Similarly, the scripts directory contains emacs.sh and screen.sh,
which get executed on the deployed node after it boots for the first
time.


Default Configuration Directory Locations
-----------------------------------------

When provision.config is first imported, it will try to load
configuration directory in ~/.provision/secrets.  If it cannot locate
one, it will then try $VIRTUAL_ENV/provision_secrets.


CHANGELOG
0.9.3 

* Remove cloudfiles dependency

0.9.2 Wed Aug 24 2011 13:09:19 GMT-0700 (PDT)

* Change from automatic to explicit template declarations
* Eliminate support for splitting scripts up by line

0.9.1 Wed Jul 27 15:36:52 PDT 2011

* Add ability to specify images as bootstrapped to change default bundles
* Order of bundles and scripts in bundles now preserved in script run order
* Add maverick and natty to image names
* Add bootstrap-python bundle
* Eliminate stack trace on normal exit
* Improve documentation on shell commands
* Give more detailed error information with exit codes
* Fix bug in processing -c flags on command line
* Support loading default provision_secrets in virtualenv directory



provision/config.py
from __future__ import print_function

"""Configuration of provision apps is a two step process.

The first step specifies which directories to use for configuring
defaults, defining resource bundles, setting API keys and which public
keys to trust. This allows users to use site-specific configuration
directories outside the source tree.

The second step specifies which bundles to install on the new node,
and any other details specific to the deployment.

Each bundle represents a set of files to be copied and scripts to be
run during node deployment."""

import argparse
import os.path
import random
import re
import string
import sys
import traceback

import socket; socket.setdefaulttimeout(600.0) # give APIs 10 minutes

import libcloud.types

import provision.collections

join = os.path.join

PROVIDERS = {
    'rackspace': libcloud.types.Provider.RACKSPACE}

IMAGE_NAMES = {
    'karmic': 'Ubuntu 9.10 (karmic)',
    'lucid': 'Ubuntu 10.04 LTS (lucid)',
    'maverick': 'Ubuntu 10.10 (maverick)',
    'natty': 'Ubuntu 11.04 (Natty)'}

DEFAULT_IMAGE_NAME = 'lucid'
DEFAULT_LOCATION_ID = 0
DEFAULT_SIZE_ID = 0

DEFAULT_PUBKEY = open(os.path.expanduser('~/.ssh/id_rsa.pub')).read()

# Note that the last directory in the path cannot start with a '.'
# due to module naming restrictions
LOCAL_DEFAULTS = os.path.expanduser('~/.provision/secrets')

VIRTUAL_ENV = os.getenv('VIRTUAL_ENV')
VIRTUAL_DEFAULTS = join(VIRTUAL_ENV, 'provision_secrets') if VIRTUAL_ENV else ''

DEFAULT_TARGETDIR = '/root/deploy'

DEFAULT_NAME_PREFIX = 'deploy-test-'

DESTROYABLE_PREFIXES = [DEFAULT_NAME_PREFIX]

TEMPLATE_RE = re.compile('#.+provision-template-type:\W*(?P<type>[\w-]+)')

TEMPLATE_TYPEMAP = {
    # http://docs.python.org/library/string.html#format-string-syntax
    'format-string': lambda text, submap: text.format(**submap),
    # http://docs.python.org/library/string.html#template-strings
    'template-string': lambda text, submap: string.Template(text).safe_substitute(submap),
    }

CODEPATH = os.path.dirname(__file__)

SCRIPTSDIR = 'scripts'
FILESDIR = 'files'
PUBKEYSDIR = 'pubkeys'

PUBKEYS = []
SUBMAP = {}
BUNDLEMAP = {}

BOOTSTRAPPED_IMAGE_NAMES = []
DEFAULT_BUNDLES = [] # these get installed if image name in BOOTSTRAPPED_IMAGE_NAMES
DEFAULT_BOOTSTRAP_BUNDLES = [] # otherwise these get installed

PATH = None

import logging
logging.basicConfig(level=logging.DEBUG,
                    format='%(asctime)s %(name)s %(levelname)s %(message)s')
logger = logging.getLogger('provision')


# error codes for corresponding exceptions
EXCEPTION = 11
MALFORMED_RESPONSE = 12
SERVICE_UNAVAILABLE = 13
DEPLOYMENT_ERROR = 14
TIMEOUT = 15

def is_node_destroyable(name, prefixes=DESTROYABLE_PREFIXES):

    """Return True if name starts with a destroyable prefix"""

    return any([name.startswith(p) for p in prefixes])


def handle_errors(callback, parsed=None, out=sys.stderr):

    """Execute the callback, optionally passing it parsed, and return
    its return value.  If an exception occurs, determine which kind it
    is, output an appropriate message, and return the corresponding
    error code."""

    try:
        if parsed:
            return callback(parsed)
        else:
            return callback()
    except libcloud.types.DeploymentError as e:
        traceback.print_exc(file=out)
        print(e, file=out)
        if hasattr(e, 'value') and hasattr(e.value, 'args') and len(e.value.args) > 0 and \
                'open_sftp_client' in e.value.args[0]:
            print('Timeout', file=out)
            return TIMEOUT
        return DEPLOYMENT_ERROR
    except libcloud.types.MalformedResponseError as e:
        traceback.print_exc(file=out)
        print(e, file=out)
        if 'Service Unavailable' in e.body:
            return SERVICE_UNAVAILABLE
        return MALFORMED_RESPONSE
    except SystemExit:
        pass
    except:
        traceback.print_exc(file=out)
        return EXCEPTION


class Bundle(object):

    """Encapsulates mappings from file and script paths on the target
    node to their local, source paths"""

    def __init__(self, scriptmap=None, filemap=None):
        """
        @type scriptmap: C{dict}
        @keyword scriptmap: Maps target path to source path for scripts

        @type filemap: C{dict}
        @keyword filemap: Maps target path to source path for files
        """
        self.scriptmap = scriptmap or provision.collections.OrderedDict()
        self.filemap = filemap or {}

def makemap(filenames, sourcedir, targetdir=None):

    """Return an OrderedDict (to preserve script run order) which maps
    filenames coming from a single local source directory to a single
    target directory.  Most useful for scripts, whose location when
    run is often unimportant, and so can all be placed in common
    directory."""

    if targetdir is None: targetdir = DEFAULT_TARGETDIR
    return provision.collections.OrderedDict(
        (join(targetdir, f),  join(sourcedir, f)) for f in filenames)

def add_bundle(name, scripts=[], files=[], scriptsdir=SCRIPTSDIR, filesdir=FILESDIR):

    """High level, simplified interface for creating a bundle which
    takes the bundle name, a list of script file names in a common
    scripts directory, and a list of absolute target file paths, of
    which the basename is also located in a common files directory.
    It converts those lists into maps and then calls new_bundle() to
    actually create the Bundle and add it to BUNDLEMAP"""

    scriptmap = makemap(scripts, join(PATH, scriptsdir))
    filemap = dict(zip(files, [join(PATH, filesdir, os.path.basename(f)) for f in files]))
    new_bundle(name, scriptmap, filemap)

def new_bundle(name, scriptmap, filemap=None):

    """Create a bundle and add to available bundles"""

    #logger.debug('new bundle %s' % name)
    if name in BUNDLEMAP:
        logger.warn('overwriting bundle %s' % name)
    BUNDLEMAP[name] = Bundle(scriptmap, filemap)

def random_str(length=6, charspace=string.ascii_lowercase+string.digits):
    return ''.join(random.sample(charspace, length))

class DictObj(object):

    """Wraps a dict so its keys are accessed like properties, using dot notation"""

    def __init__(self, d):
        self.__dict__['d'] = d

    def __setattr__(self, key, value):
        if self.__dict__['d'].get(key):
            logger.warn('overwriting config.{0}'.format(key))
        self.__dict__['d'][key] = value

    def __getattr__(self, key):
        return self.__dict__['d'][key]


def import_by_path(path):

    """Append the path to sys.path, then attempt to import module with
    path's basename, finally making certain to remove appended path.

    http://stackoverflow.com/questions/1096216/override-namespace-in-python"""

    sys.path.append(os.path.dirname(path))
    try:
        return __import__(os.path.basename(path))
    except ImportError:
        logger.warn('unable to import {0}'.format(path))
    finally:
        del sys.path[-1]

def init_module(path):

    """Attempt to import a Python module located at path.  If
    successful, and if the newly imported module has an init()
    function, then set the global PATH in order to simplify the
    add_bundle() interface and call init() on the module, passing the
    current global namespace, conveniently converted into a DictObj so
    that it can be accessed with normal module style dot notation
    instead of as a dict.

    http://stackoverflow.com/questions/990422/how-to-get-a-reference-to-current-modules-attributes-in-python"""

    mod = import_by_path(path)
    if mod is not None and hasattr(mod, 'init'):
        logger.debug('calling init on {0}'.format(mod))
        global PATH
        PATH = path
        mod.init(DictObj(globals()))

def load_pubkeys(loadpath, pubkeys):

    """Append the file contents in loadpath directory onto pubkeys list"""

    filenames = os.listdir(loadpath)
    logger.debug('loading authorized pubkeys {0}'.format(filenames))
    for filename in filenames:
        pubkeys.append(open(join(loadpath, filename)).read())

def normalize_path(path, relative_to=os.getcwd()):

    """Return normalized path.  If path is not user-expandable or
    absolute, treat it relative to relative_to"""

    path = os.path.expanduser(os.path.normpath(path))

    if os.path.isabs(path):
        return path
    else:
        return join(relative_to, path)

def configure(paths, relative_to):

    """Iterate on each configuration path, collecting all public keys
    destined for the new node's root account's authorized keys.
    Additionally attempt to import path as python module."""

    if not paths:
        return
    for path in [normalize_path(p, relative_to) for p in paths]:
        logger.debug('configuration path {0}'.format(path))
        pubkeys_path = join(path, PUBKEYSDIR)
        if os.path.exists(pubkeys_path):
            load_pubkeys(pubkeys_path, PUBKEYS)
        init_module(path)

def parser():

    """Return a parser for setting one or more configuration paths"""

    parser = argparse.ArgumentParser()
    parser.add_argument('-c', '--config_paths', default=[], action='append',
                        help='path to a configuration directory')
    return parser

def add_auth_args(parser, config):

    """Return a parser for configuring authentication parameters"""

    parser.add_argument('-p', '--provider', default=config.DEFAULT_PROVIDER)
    parser.add_argument('-u', '--userid', default=config.DEFAULT_USERID)
    parser.add_argument('-k', '--secret_key', default=config.DEFAULT_SECRET_KEY)
    return parser

def reconfig(main_parser, args=sys.argv[1:]):

    """Parse any config paths and reconfigure defaults with them
    http://docs.python.org/library/argparse.html#partial-parsing
    Return parsed remaining arguments"""

    parsed, remaining_args = parser().parse_known_args(args)
    configure(parsed.config_paths, os.getcwd())
    return main_parser().parse_args(remaining_args)

defaults = ['defaults']
if os.path.exists(LOCAL_DEFAULTS):
    defaults.append(LOCAL_DEFAULTS)
if os.path.exists(VIRTUAL_DEFAULTS):
    defaults.append(VIRTUAL_DEFAULTS)
configure(defaults, CODEPATH)

